applications/nrf5340_audio/src/audio_module/amod_api.c

```diff
@@ -0,0 +1,1136 @@
+		return ret;
+	}
+
+	/* fill data */
+	data_in_msg->object = object;
+	data_in_msg->tx_handle = (struct amod_handle *)tx_handle;
+	data_in_msg->response_cb = data_in_response_cb;
+
+	ret = data_fifo_block_lock(&rx_handle->in_msg, (void **)&data_in_msg,
+				   sizeof(struct _amod_in_message));
+	if (ret) {
+		data_fifo_block_free(&rx_handle->in_msg, (void **)&data_in_msg);
+
+		LOG_DBG("Module %s failed to return output buffer, ret %d", rx_handle->name, ret);
+		return ret;
+	}
+
+	LOG_DBG("Data sent to module %s", rx_handle->name);
+
+	return 0;
+}
+
+/**
+ * @brief Function to clean up items if a procedure fails.
+ *
+ */
+static void clean_up(struct _amod_handle *hdl, struct _amod_in_message **in_msg,
+		     struct _amod_out_message **out_msg, char **data)
+{
+	if ((*in_msg)->response_cb != NULL) {
+		(*in_msg)->response_cb((*in_msg)->tx_handle, (*in_msg)->object);
+	}
+
+	if ((*in_msg) != NULL) {
+		data_fifo_block_free(&hdl->in_msg, (void **)(*in_msg));
+	}
+
+	if ((*out_msg) != NULL) {
+		data_fifo_block_free(&hdl->out_msg, (void **)(*out_msg));
+	}
+
+	if ((*data) != NULL) {
+		k_mem_slab_free(&hdl->data_slab, (void **)(*data));
+	}
+}
+
+/**
+ * @brief The thread that receives data from outside the system and passes it into the audio system.
+ *
+ * @param handle  The handle for this modules instance
+ *
+ * @return 0 if successful, error value
+ */
+static int module_thread_input(struct amod_handle *handle)
+{
+	struct _amod_handle *hdl = (struct _amod_handle *)handle;
+	struct _amod_handle *hdl_to;
+	struct _amod_in_message *in_msg = NULL;
+	struct _amod_out_message *out_msg;
+	int ret;
+	char *data;
+
+	if (handle == NULL) {
+		LOG_DBG("Error in the module task as handle NULL");
+		return -EINVAL;
+	}
+
+	/* Execute thread */
+	for (;;) {
+		out_msg = NULL;
+		data = NULL;
+
+		if (hdl->functions->data_process != NULL) {
+			/* Get a new output buffer */
+			ret = k_mem_slab_alloc(&hdl->data_slab, (void **)&data, K_FOREVER);
+			if (ret) {
+				clean_up(hdl, &in_msg, &out_msg, &data);
+
+				LOG_DBG("No free data buffer for module %st, ret %d", hdl->name,
+					ret);
+				continue;
+			}
+
+			ret = data_fifo_pointer_first_vacant_get(&hdl->out_msg, (void **)&out_msg,
+								 K_FOREVER);
+			if (ret) {
+				clean_up(hdl, &in_msg, &out_msg, &data);
+
+				LOG_DBG("No free message for module %s, ret %d", hdl->name, ret);
+				continue;
+			}
+
+			/* Configure new audio object */
+			out_msg->object.data = data;
+			out_msg->object.data_size = hdl->data_size;
+
+			/* Process the input audio object */
+			ret = hdl->functions->data_process(handle, NULL, &out_msg->object);
+			if (ret) {
+				clean_up(hdl, &in_msg, &out_msg, &data);
+
+				LOG_DBG("Data process error in module %s, ret %d", hdl->name, ret);
+				continue;
+			}
+
+			/* Send input audio object to next module(s) */
+			if (!sys_slist_is_empty(&hdl->hdl_dest_list)) {
+				k_mutex_lock(&hdl->dest_mutex, K_FOREVER);
```

Comment:
That is a very good point, but this is only to prevent the connect or disconnect jumping in while we are sending objects to the next module. The data_send() queues the object for the next module and this is a no wait with error if no free message, as it will be time critical this data moves down the stream in a timely fashion. I did have a number of timeouts for the various message queue actions, but convinced myself we didn't need them. We should discuss this when you have time.